src/cats/CatSlideshow.tsx

```diff
@@ -0,0 +1,178 @@
+
+type CatAPISearchResponseItemType = {
+  id: string;
+  url: string;
+};
+
+type CatAPISearchResponseType = Array<CatAPISearchResponseItemType>;
+type SlideshowPageType = Array<{ id: string; url: string }>;
+type SlideshowPagesType = Record<number, PageType<SlideshowPageType, string>>;
+
+const PAGE_SIZE = 20;
+const PAGE_PREFETCH_LIMIT = 8;
+
+function fetchPageFromCatsApi(
+  pageIndex: number,
+  selectedBreedID: string,
+  fetchPage: FetchPageCallbackType<SlideshowPageType>
+) {
+  const { url, options } = getCatsApiFetchParams("/images/search", {
+    breed_id: selectedBreedID === "all" ? "" : selectedBreedID,
+    limit: String(PAGE_SIZE),
+    order: "ASC",
+    page: String(pageIndex),
+  });
+
+  fetchPage({
+    url,
+    options,
+    index: pageIndex,
+    key: selectedBreedID,
+    getPageData: async (response): Promise<SlideshowPageType> => {
+      const json: CatAPISearchResponseType = await response.json();
+      return json.map((item) => ({ id: item.id, url: item.url }));
+    },
+    getMetadata: (response): Record<string, string> => {
+      const count = response.headers.get("pagination-count");
+      return count !== null ? { paginationCount: count } : {};
+    },
+  });
+}
+
+function getPageIndex(index: number) {
+  return Math.floor(index / PAGE_SIZE);
+}
+
+function getImageURL(pages: SlideshowPagesType, index: number) {
+  const pageIndex = getPageIndex(index);
+  const offset = index % PAGE_SIZE;
+  return pages?.[pageIndex]?.data?.[offset]?.url;
+}
+
+export default function CatSlideshow({
+  selectedBreedID,
+}: CatSlideshowPropsType) {
+  const [direction, setDirection] = useState(Direction.Next);
+  const [index, setIndex] = useState<number>(0);
+  const [visibleIndex, setVisibleIndex] = useState<number | null>(null);
+  const [maxIndex, setMaxIndex] = useState<number | null>(null);
+  const previousBreedID = usePrevious<string>(selectedBreedID);
+  const { pages, metadata, status, fetchPage, resetPages } = usePaginatedFetch<
+    SlideshowPageType,
+    string
+  >(selectedBreedID, "loading");
+  const { imageLoadMap, resetPrefetch } = usePrefetchImages(
+    index,
+    useCallback((i) => getImageURL(pages, i), [pages])
+  );
+
+  useEffect(() => {
+    if (selectedBreedID !== previousBreedID) {
+      resetPages(selectedBreedID);
+      resetPrefetch();
+      setIndex(0);
+      setDirection(Direction.Next);
+      setVisibleIndex(null);
+      setMaxIndex(null);
+    }
+  }, [selectedBreedID, resetPages, resetPrefetch, previousBreedID]);
+
+  useEffect(() => {
+    const pageIndex = getPageIndex(index);
+    const prefetchPageIndex = getPageIndex(index + PAGE_PREFETCH_LIMIT);
+    const doFetch = (i: number) =>
+      fetchPageFromCatsApi(i, selectedBreedID, fetchPage);
+
+    if (!pages[pageIndex]) {
+      doFetch(pageIndex);
+    }
+
+    if (prefetchPageIndex !== pageIndex && !pages[prefetchPageIndex]) {
+      doFetch(prefetchPageIndex);
+    }
+
+    if (metadata?.paginationCount !== undefined) {
```

Comment:
this should be in its own effect -- or ideally in the reducer itself

Reply:
This was originally it's own effect, but I got some feedback from @chadwhiting that this component had too many side effects. My understanding was that for separation of concerns and performance reasons, it's best to use as many separate effects as you need.

I think the real win for this component was breaking the image prefetching out into it's own hook. It is a clean separation of functionality and has real reusability.